### PR TITLE
Use FreeIPA API to authenticate and authorize users instead of LDAP.

### PR DESCRIPTION
**Pull request by felddy**
# 

## 🗣 Description

- Replace LDAP calls with kerberized calls to the FreeIPA API.
- Compare user TLS certificates instead of just their CN.
- Modernize role.
- Upstream changes.

## 💭 Motivation and Context

We were running into issues with LDAP comparing DNs.  Specifically multi-value RDNs.  The order of AVAs with and RDNs is insignificant (is a `set`).  There is no way to "normalize" a CN that will thwart our badging offices ability to scramble up our UIDs nicely.  

See: https://frasertweedale.github.io/blog-redhat/posts/2019-05-28-a-dn-is-not-a-string.html

We are now configuring OpenVPN to pass the user's entire public certificate to us as a file.  It is then sent to FreeIPA for matching.

As an added benefit we can now ask FreeIPA interesting questions not only about group membership, but multiple attributes that would deactivate an account, preventing login. 

## 🧪 Testing

Tested in staging.
Molecule tests.

## 🚥 Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (causes existing functionality to change)

## ✅ Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] Big thanks to @jsf9k for fixing things I could not.

### CODE REVIEW

File: vars/Debian.yml [hunk 1]
```diff
@@ -20,3 +19,14 @@ iptables_persistence_package_names:
...
+# The packages to install to support the common name verification
+# script
+verify_cn_package_names:
+  - libkrb5-dev
+
+# The pip packages to install to support the common name verification
+# script
+verify_cn_pip_package_names:
+  - python-freeipa
+  - requests-gssapi
```
Comment (jsf9k) [inline vars/Debian.yml:32]: I still question whether requests-gssapi is required for any distribution.  I don't see it imported or used anywhere.

**Review by jsf9k — CHANGES_REQUESTED** [review]
I will approve once you have addressed the concern I have raised before.  Please see my comment.

File: files/verify-cn.py [hunk 2]
```diff
@@ -19,94 +20,152 @@
...
-    )
-    if len(rec) == 0:
-        logging.warning("No matching certificate found in LDAP")
-        logging.warning(f"Searched for: {dn}")
+def query_freeipa(client_certificate: str, realm: str, group: str) -> bool:
+    """Determine if a user should be allowed to connect."""
+    # Create the FreeIPA client
+    ipa_client: ClientMeta = ClientMeta(dns_discovery=realm)
+
+    # Login client in using kerberos credentials
```
Comment (dav3r) [inline files/verify-cn.py]: There's an extra `"in"` in this comment- pick one to keep and one to toss.

File: files/verify-cn.py [hunk 3]
```diff
@@ -19,94 +20,152 @@
...
+
+    if response["count"] == 0:
+        logging.warning("No matching user found.")
         return False
 
-    if len(rec) > 1:
-        logging.warning(f"More than 1 record found in LDAP.  Count: {len(rec)}.")
+    if response["count"] > 1:
+        logging.warning(
+            f"More than 1 user matched certificate.  Count: {response['count']}."
```
Comment (dav3r) [inline files/verify-cn.py]: Can we log all of the UIDs found in this case?  That would make resolving this situation much easier.

File: files/verify-cn.py [hunk 4]
```diff
@@ -19,94 +20,152 @@
...
+        logging.warning(f"{matched_uid} account IS PRESERVED.")
+        account_not_preserved = False
 
-def main():
+    # Pass judgement on the user
+    if group_ok and account_enabled and account_not_preserved:
+        logging.info(f"{matched_uid} will be permitted access.")
+        return True
+    else:
+        logging.warning(f"{matched_uid} ACCESS DENIED.")
```
Comment (dav3r) [inline files/verify-cn.py]: ⛔⛔⛔ 👍

File: files/verify-cn.yml [hunk 5]
```diff
@@ -1,3 +1,4 @@
 ---
-realm: COOL.CYBER.DHS.GOV
+log_level: INFO
+realm: STAGING.COOL.CYBER.DHS.GOV
```
Comment (dav3r) [inline files/verify-cn.yml:3]: Is this `STAGING.` supposed to be here?

**Review by dav3r — COMMENTED** [review]
In addition to @jsf9k's important question, I added a few minor comments/questions.

File: files/verify-cn.py (hunk 4 above)
Comment (jsf9k) [reply files/verify-cn.py]: Previously @mcdonnnj pointed out that there are good reasons to prefer format strings to f-strings in `logging` calls like this.  Not sure if you had forgotten this or were purposefully disregarding it.  We should as a team choose one style and use it everywhere.

Comment (dav3r) [reply files/verify-cn.py]: Forgot about that- we should avoid f-strings in logging statements.

https://blog.pilosus.org/posts/2020/01/24/python-f-strings-in-logging/

File: files/verify-cn.py [hunk 6]
```diff
@@ -19,94 +20,152 @@
...
-def kinit():
+def load_client_certificate() -> Optional[str]:
+    """Read certificate data from file identified by environment variable."""
+    # OpenVPN sends us the client's certificate though the peer_cert environment
+    # variable.
+    try:
+        client_certificate_path: Path = Path(os.environ[PEER_CERT_VARIABLE])
+    except KeyError:
+        logging.critical(
+            "OpenVPN did not set the 'peer_cert' environement variable."
```
Comment (dav3r) [inline files/verify-cn.py]: Typo: `environement`

File: vars/Debian.yml (hunk 1 above)
Comment (felddy) [reply vars/Debian.yml:32]: This is required to use the `login_kerberos` function of `python-freeipa`.
See: https://python-freeipa.readthedocs.io/en/latest/#python_freeipa.client.Client.login_kerberos

File: files/verify-cn.yml (hunk 5 above)
Comment (felddy) [reply files/verify-cn.yml:3]: yes. This file is just a placeholder.  The real file is created by cloud-init.

File: vars/Debian.yml (hunk 1 above)
Comment (jsf9k) [reply vars/Debian.yml:32]: Got it!

**Review by jsf9k — APPROVED** [review]

File: files/primary.conf [hunk 7]
```diff
@@ -313,6 +313,13 @@ verb 3
 # can automatically reconnect.
 explicit-exit-notify 1
 
+# Store the certificates the clients uses upon connection to this directory.
```
Comment (jsf9k) [inline files/primary.conf]: "clients uses" is _poor_ English.

File: files/verify-cn.py (hunk 6 above)
Comment (dav3r) [reply files/verify-cn.py]: Fixed in https://github.com/cisagov/ansible-role-openvpn/pull/18/commits/79c9a9199c6ee6436362e3181db86c6c46377c22 👍

File: files/verify-cn.py (hunk 2 above)
Comment (dav3r) [reply files/verify-cn.py]: Fixed in https://github.com/cisagov/ansible-role-openvpn/pull/18/commits/32f83bde13b8778d07ca55771f7a668de2093ef1 👍

File: files/primary.conf (hunk 7 above)
Comment (felddy) [reply files/primary.conf]: My bad.  Copied from man page without proofreading it.

File: files/verify-cn.py (hunk 3 above)
Comment (felddy) [reply files/verify-cn.py]: Good idea. 
See: c065295

File: files/verify-cn.py (hunk 4 above)
Comment (dav3r) [reply files/verify-cn.py]: Begrudgingly adjusted by @felddy in https://github.com/cisagov/ansible-role-openvpn/pull/18/commits/3b0a4f629408e4c26f5227eee74117e2705f3415. 😜

**Review by jsf9k — APPROVED** [review]
:zap:

**Review by dav3r — APPROVED** [review]
This is some 🌶️ 💩 right here!  🚀

File: files/verify-cn.py (hunk 3 above)
Comment (dav3r) [reply files/verify-cn.py]: The future admin who encounters this situation thanks you sir!  🍹

File: files/verify-cn.py (hunk 4 above)
Comment (felddy) [reply files/verify-cn.py]: 😏

**Pull request merged**